futures-util/src/async_await/join.rs

```diff
@@ -94,30 +92,34 @@ macro_rules! try_join {
             let mut $fut = $crate::future::maybe_done($fut);
             $crate::pin_mut!($fut);
         )*
-        let res: $crate::core_reexport::result::Result<(), _> = loop {
+
+        let res: $crate::core_reexport::result::Result<_, _> = await!($crate::future::poll_fn(move |cx| {
             let mut all_done = true;
             $(
-                if $crate::poll!($fut.reborrow()).is_pending() {
+                if $crate::future::Future::poll($fut.reborrow(), cx).is_pending() {
                     all_done = false;
                 } else if $fut.reborrow().output_mut().unwrap().is_err() {
                     // `.err().unwrap()` rather than `.unwrap_err()` so that we don't introduce
                     // a `T: Debug` bound.
-                    break $crate::core_reexport::result::Result::Err(
-                        $fut.reborrow().take_output().unwrap().err().unwrap()
+                    return $crate::core_reexport::task::Poll::Ready(
+                        $crate::core_reexport::result::Result::Err(
+                            $fut.reborrow().take_output().unwrap().err().unwrap()
+                        )
                     );
                 }
             )*
             if all_done {
-                break $crate::core_reexport::result::Result::Ok(());
+                $crate::core_reexport::task::Poll::Ready(
+                    $crate::core_reexport::result::Result::Ok(($(
+                    // `.ok().unwrap()` rather than `.unwrap()` so that we don't introduce
+                    // an `E: Debug` bound.
+                    $fut.reborrow().take_output().unwrap().ok().unwrap(),
+                )*)))
             } else {
-                $crate::pending!();
+                $crate::core_reexport::task::Poll::Pending
             }
-        };
+        }));
 
-        res.map(|()| ($(
-            // `.ok().unwrap()` rather than `.unwrap()` so that we don't introduce
-            // an `E: Debug` bound.
-            $fut.reborrow().take_output().unwrap().ok().unwrap(),
-        )*))
+        res
```

Comment:
It's basically:
```rust
let res = /* ... */;
res
```
You could return the value from the expression above directly

Reply:
It's intentionally done this way in order to provide the `: Result<_, _>` annotation .